pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,465 @@
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Cluster()
+	clusterObjKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("CR '%s' not found: %w", clusterName, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch CR '%s': %w", clusterName, err)
+	}
+
+	user := consts.DefaultOperatorUser
+	pass, err := users.PasswordFromSecret(user, infinispan.GetSecretName(), namespace, z.Kube)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	httpConfig := ispnclient.HttpConfig{
+		Username:  user,
+		Password:  pass,
+		Namespace: namespace,
+		Protocol:  strings.ToLower(string(infinispan.GetEndpointScheme())),
+	}
+	httpClient := curl.New(httpConfig, z.Kube)
+
+	if phase == ZeroInitialized {
+		return z.execute(httpClient, request, instance)
+	}
+	// Phase must be ZeroRunning, so wait for execution to complete
+	return z.waitForExecutionToComplete(httpClient, request, instance)
+}
+
+func (z *Controller) initializeResources(request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	ctx := context.Background()
+	name := request.Name
+	namespace := request.Namespace
+	clusterName := instance.Cluster()
+	clusterKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+
+	infinispan := &v1.Infinispan{}
+	if err := z.Client.Get(ctx, clusterKey, infinispan); err != nil {
+		z.Log.Info(fmt.Sprintf("Unable to load Infinispan Cluster '%s': %w", clusterName, err))
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if err := ensureClusterStability(infinispan); err != nil {
+		return reconcile.Result{}, fmt.Errorf("Infinispan not stable: %w", err)
+	}
+
+	statefulset := &appsv1.StatefulSet{}
+	if err := z.Client.Get(ctx, clusterKey, statefulset); err != nil {
+		z.Log.Info(fmt.Sprintf("Unable to load Infinispan StatefulSet '%s': %w", clusterName, err))
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	spec, err := instance.Init()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	configMap, err := z.configureZeroCapacity(name, namespace, spec, infinispan, instance)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to create zero-capacity configuration: %w", err)
+	}
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    spec.PodLabels,
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, z.Client, pod, func() error {
```

Comment:
@ryanemerson not sure that you need to update zero pod. There are a few fields in the Pod object that could be updated

Reply:
How can I update these fields without updating the pod itself? Also note, that in most cases the pod will actually be created here.

Reply:
I'm not sure, that following fields need to be updated in our case `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations`
We just need to create a pod if it not yet present. Nothing more.

Reply:
That's what this method does. The pod doesn't exist, so it's created and the fields populated:

```
// CreateOrUpdate creates or updates the given object in the Kubernetes
// cluster. The object's desired state must be reconciled with the existing
// state inside the passed in callback MutateFn.
//
// The MutateFn is called regardless of creating or updating an object.
```

Reply:
Yes, true, but you need define in MutateFun is it new object or not and which fields to create or to update 
https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil